palace/drivers/basesolver.cpp

```diff
@@ -84,6 +89,282 @@ BaseSolver::BaseSolver(const IoData &iodata, bool root, int size, int num_thread
   }
 }
 
+std::string BaseSolver::IterationPostDir(int iter) const
+{
+  std::string dir = fmt::format("{}Adapt{:0>3d}/", post_dir, iter);
+  // Create directory for output.
+  if (root && !std::filesystem::exists(dir))
+  {
+    std::filesystem::create_directories(dir);
+  }
+  return dir;
+}
+
+namespace
+{
+
+// Helper function that returns an array of indices corresponding to marked elements.
+mfem::Array<int> MarkedElements(double threshold, const Vector &e)
+{
+  mfem::Array<int> ind;
+  ind.Reserve(e.Size());
+  for (int i = 0; i < e.Size(); i++)
+  {
+    if (e[i] >= threshold)
+    {
+      ind.Append(i);
+    }
+  }
+  return ind;
+}
+
+// Helper function responsible for rebalancing the mesh, and optionally writing meshes from
+// the intermediate stages to disk.
+void RebalanceMesh(std::unique_ptr<mfem::ParMesh> &mesh, const IoData &iodata,
```

Comment:
Seems to me like this code belongs in `geodata.cpp`, next to `RebalanceConformalMesh`?